backend/service/chaos/experimentation/experimentstore/experiment_specification.go

```diff
@@ -0,0 +1,91 @@
+package experimentstore
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/golang/protobuf/ptypes/timestamp"
+
+	experimentation "github.com/lyft/clutch/backend/api/chaos/experimentation/v1"
+	"github.com/lyft/clutch/backend/id"
+)
+
+const runIdRegExp = `^[A-Za-z0-9-._~]+$`
+
+type ExperimentSpecification struct {
+	RunId     string
+	ConfigId  string
+	StartTime time.Time
+	EndTime   *time.Time
+	Config    *any.Any
+}
+
+// New returns a new experimentationSpecification instance.
+func NewExperimentSpecification(ced *experimentation.CreateExperimentData, now time.Time) (*ExperimentSpecification, error) {
+	var runId = ced.RunId
+	// If the run identifier is not provided, generate one
+	if runId == "" {
+		runId = id.NewID().String()
+	} else if !regexp.MustCompile(runIdRegExp).MatchString(runId) {
+		return nil, fmt.Errorf("provided experiment runId (%v) contained unallowed characters and was not matched by \"%v\" regular expresion", runId, runIdRegExp)
+	}
+
+	if ced.Config == nil {
+		return nil, errors.New("experiment config cannot be equal to nil")
+	}
+
+	// If the start time is not provided, default to current time
+	startTime := now
+	if ced.StartTime != nil {
+		startTime = ced.StartTime.AsTime()
+	}
+
+	if now.After(startTime) {
+		return nil, fmt.Errorf("current time (%v) must be equal to or after experiment start time (%v)", now, startTime)
+	}
```

Comment:
Re phrasing - good point, I'm going to update the error message.

When it comes to `startTime: time.Now()`. side note: I will improve the proto documentation. I'm not worried about startTime: time.Now() tho because if a user wants to start experiment immediately after a call succeeds they should just omit passing `startTime` argument and allow the server to default its value to `time.Now()` then (this is not clear now since the documentation does not mention this).

Reply:
what do u think?